src/v/pandaproxy/kafka_client_cache.cc

```diff
@@ -0,0 +1,185 @@
+
+kafka_client_cache::kafka_client_cache(
+  YAML::Node const& cfg, size_t max_size, std::chrono::milliseconds keep_alive)
+  : _config{cfg}
+  , _cache_max_size{max_size}
+  , _keep_alive{keep_alive} {}
+
+client_ptr kafka_client_cache::make_client(
+  credential_t user, config::rest_authn_method authn_method) {
+    kafka::client::configuration cfg{
+      to_yaml(_config, config::redact_secrets::no)};
+
+    // Set the principal when the request is using HTTP Basic AuthN
+    if (authn_method == config::rest_authn_method::http_basic) {
+        // Need to specify type or else bad any_cast runtime error
+        cfg.sasl_mechanism.set_value(ss::sstring{"SCRAM-SHA-256"});
+        cfg.scram_username.set_value(std::move(user.name));
+        cfg.scram_password.set_value(std::move(user.pass));
+    }
+
+    return ss::make_lw_shared<kafka::client::client>(
+      to_yaml(cfg, config::redact_secrets::no));
+}
+
+ss::future<client_ptr> kafka_client_cache::fetch_or_insert(
+  credential_t user, config::rest_authn_method authn_method) {
+    auto& inner_list = _cache.get<underlying_list>();
+    auto& inner_hash = _cache.get<underlying_hash>();
+    ss::sstring k{user.name};
+    auto it_hash = inner_hash.find(k);
+
+    client_ptr client;
+
+    // When no client is found ...
+    if (it_hash == inner_hash.end()) {
+        if (_cache.size() >= _cache_max_size) {
+            if (!_cache.empty()) {
+                auto item = inner_list.back();
+                vlog(plog.debug, "Cache size reached, evicting {}", item.key);
+                inner_list.pop_back();
+                _evicted_items.push_back(item);
+            }
+        }
+
+        vlog(plog.debug, "Make client for user {}", k);
+
+        client = make_client(user, authn_method);
+        inner_list.emplace_front(k, client);
+    } else {
+        // Make a new client if the passwords don't match
+        if (it_hash->client->config().scram_password.value() != user.pass) {
+            vlog(plog.debug, "Password changed. Remake client for user {}", k);
+            client_ptr old = it_hash->client;
+            client_ptr next = make_client(user, authn_method);
+            co_await _lock.with([k, next] {
+                // Try connect as a safety check
+                return next->connect().handle_exception(
+                  [k](std::exception_ptr ex) {
+                      // The failed connection is dropped
+                      vlog(
+                        plog.debug,
+                        "Connection for {} already formed. Dropping dup: {}",
+                        k,
+                        ex);
+                  });
+            });
+
+            inner_hash.modify(
+              it_hash, [next](timestamped_user& item) { item.client = next; });
+
+            // Add a prefix to the evicted key so debugging logs
+            // is easier
+            _evicted_items.emplace_back(
+              k + random_generators::gen_alphanum_string(8), old);
```

Comment:
This is tricky, but I think I have an optimisation and simplification.

* The password has already been validated against the user, so we know it's good.
* The password is associated with the connection.
* Changing the password doesn't break the connection.
* The next connect will need the new password.

```suggestion
        // If the passwords don't match, update the password on the client, so
        // that it can reconnect.
        if (it_hash->client->config().scram_password.value() != user.pass) {
            vlog(plog.debug, "Updating password for user {}", k);
            it_hash->client->config().scram_password.set_value(user.pass);
```

The function no longer needs to be asynchronous.

Reply:
Excellent point. I'll make the change